### PR TITLE
fix(notifications): set a default time for lastNotifiedMessageDate for convresations

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Having the default null value for lastNotifiedMessageDate in conversations is causing some problem with some edge cases.
We set here the current time when we insert the conversation in the database as the lastNotifiedMessageDate.
### Issues

Some edge cases bugs on notifications.

### Solutions

Setting a default value.

### Dependencies (Optional)

Needs to update tests cases on AR to pass with new non-null enforcement.

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt [hunk 1]
```diff
@@ -226,7 +227,7 @@ internal class ConnectionDataSource(
                         teamId = null,
                         protocolInfo = ConversationEntity.ProtocolInfo.Proteus,
                         creatorId = connection.from,
-                        lastNotificationDate = null,
+                        lastNotificationDate = Clock.System.now().toString(),
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt:230]: Please do not use regular `.toString()`, it omits milliseconds in the string if they are equal to 0, that's why I had to do this change: https://github.com/wireapp/kalium/pull/1258

Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt:230]: Awesome, thanks. So I'll wait for you RP to get merged and then I'll update my PR.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt [hunk 2]
```diff
@@ -226,7 +227,7 @@ internal class ConnectionDataSource(
                         teamId = null,
                         protocolInfo = ConversationEntity.ProtocolInfo.Proteus,
                         creatorId = connection.from,
-                        lastNotificationDate = null,
+                        lastNotificationDate = DateTimeUtil.currentIsoDateTimeString(),
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt:230]: Now I'm not sure, won't that create similar issue as the one Vitor debugged last night? For connections maybe we can use `lastUpdate` field?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationGroupRepository.kt [hunk 3]
```diff
@@ -68,7 +69,10 @@ internal class ConversationGroupRepositoryImpl(
             .flatMap { conversationResponse ->
                 val teamId = selfUser.teamId
                 val conversationEntity = conversationMapper.fromApiModelToDaoModel(
-                    conversationResponse, mlsGroupState = ConversationEntity.GroupState.PENDING_CREATION, teamId
+                    conversationResponse,
+                    mlsGroupState = ConversationEntity.GroupState.PENDING_CREATION,
+                    teamId,
+                    DateTimeUtil.currentIsoDateTimeString()
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationGroupRepository.kt:75]: Same here, we should use date-time from the server, here for conversations maybe we can use `lastEventTime`?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt [hunk 4]
```diff
@@ -253,6 +254,7 @@ internal class ConversationDataSource internal constructor(
                         )
                     },
                     selfTeamIdProvider().getOrNull(),
+                    lastNotificationDate = DateTimeUtil.currentIsoDateTimeString()
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt:257]: Maybe we shouldn't set it here at all, because we use that `persistConversations` when syncing, right? So during every sync we will update this `lastNotificationDate` to the current device time and won't get notified about messages that happened before that time, or maybe I'm missing something 🤔

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/Conversation.kt [hunk 5]
```diff
@@ -182,7 +183,7 @@ sealed class ConversationDetails(open val conversation: Conversation) {
             protocol = protocolInfo,
             mutedStatus = MutedConversationStatus.AllAllowed,
             removedBy = null,
-            lastNotificationDate = null,
+            lastNotificationDate = DateTimeUtil.currentIsoDateTimeString(),
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/Conversation.kt:186]: We shouldn't rely on local device's time, maybe for connection it should be `lastModifiedDate` or even just don't set a default value at all 🤔

**Review by saleniuk — COMMENTED** [review]
I think @vitorhugods should take a look as he's working on a fix for the notification timing issue, but I think we shouldn't use the device's local time, maybe for each case we can find a proper date-time field that's returned by server?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt (hunk 2 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/connection/ConnectionRepository.kt:230]: we don't have access to that field! this is for the conversation that we don't have anything for it and instead of null we set the current time to it.

**Comment by vitorhugods** [comment]
> I think @vitorhugods should take a look as he's working on a fix for the notification timing issue, but I think we shouldn't use the device's local time, maybe for each case we can find a proper date-time field that's returned by server?

I guess we could just use `Instant.DISTANT_PAST` as well. This way we avoid unnecessary null checks?
